native/dlaqps.go

```diff
@@ -0,0 +1,208 @@
+// Dlaqps is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlaqps(m, n, offset, nb int, a []float64, lda int, jpvt []int, tau, vn1, vn2, auxv, f []float64, ldf int) (kb int) {
+	checkMatrix(m, n, a, lda)
+	checkMatrix(nb, nb, f, ldf)
+	if len(jpvt) != n {
+		panic(badIpiv)
+	}
+	if len(tau) < n {
+		panic(badTau)
+	}
+	if len(vn1) < n {
+		panic(badVn1)
+	}
+	if len(vn2) < n {
+		panic(badVn2)
+	}
+	if len(auxv) < nb {
+		panic(badAuxv)
+	}
+
+	lastrk := min(m, n+offset)
+	lsticc := 0
+	k := 0
+	tol3z := math.Sqrt(dlamchE)
+
+	bi := blas64.Implementation()
+
+	var rk int
+	for k < nb && lsticc == 0 {
+		k++
+		rk = offset + k
+
+		// Determine kth pivot column and swap if necessary.
+		p := k + bi.Idamax(n-k, vn1[k:], 1)
+		if p != k {
+			bi.Dswap(m, a[p:], lda, a[k:], lda)
+			bi.Dswap(k, f[p*ldf:], 1, f[k*ldf:], 1)
+			jpvt[p], jpvt[k] = jpvt[k], jpvt[p]
+			vn1[p] = vn1[k]
+			vn2[p] = vn2[k]
+		}
+
+		// Apply previous Householder reflectors to column K:
+		//
+		// A[rk:m, k] = A[rk:m, k] - A[rk:m, 0:k-1]*F[k, 0:k-1]**T.
+		if k > 0 {
+			bi.Dgemv(blas.NoTrans, m-rk, k-1, -1,
+				a[rk*lda:], lda,
+				f[k*ldf:], ldf,
+				1,
+				a[rk*lda+k:], 1)
+		}
+
+		// Generate elementary reflector H[k].
+		if rk < m {
+			_, tau[k] = impl.Dlarfg(m-rk, a[rk*lda+k], a[rk*lda+k:], lda)
+		} else {
+			_, tau[k] = impl.Dlarfg(1, a[rk*lda+k], a[rk*lda+k:], lda)
+		}
+
+		akk := a[rk*lda+k]
+		a[rk*lda+k] = 1
+
+		// Compute kth column of F:
+		//
+		// Compute F[k+1:n, k] = tau[k]*A[rk:m, k+1:n]**T*A[rk:m, k].
+		if k < n {
+			bi.Dgemv(blas.Trans, m-rk, n-k, tau[k],
+				a[rk*lda+k+1:], lda,
+				a[rk*lda+k:], lda,
+				0,
+				f[(k+1)*ldf+k:], ldf)
+		}
+
+		// Padding F[0:k, k] with zeros.
+		for j := 0; j < k; j++ {
+			f[j*ldf+k] = 0
+		}
+
+		// Incremental updating of F:
+		//
+		// F[0:n, k] := F[0:n, k] - tau[k]*F[0:n, 0:k-1]*A[rk:m, 0:k-1]**T*A[rk:m,k].
+		if k > 0 {
+			bi.Dgemv(blas.Trans, m-rk, k-1, -tau[k],
+				a[rk*lda:], lda,
+				a[rk*lda+k:], lda,
+				0,
+				auxv, 1)
+			bi.Dgemv(blas.NoTrans, n, k-1, 1,
+				f, ldf,
+				auxv, 1,
+				1,
+				f[k:], ldf)
+		}
+
+		// Update the current row of A:
+		//
+		// A[rk, k+1:n] = A[rk, k+1:n] - A[rk, 0:k]*F[k+1:n, 0:k]**T.
+		if k < n {
+			bi.Dgemv(blas.NoTrans, n-k, k, -1,
+				f[(k+1)*ldf:], ldf,
+				a[rk*lda:], lda,
+				1,
+				a[rk*lda+k+1:], 1)
+		}
+
+		// Update partial column norms.
+		if rk < lastrk {
+			for j := k + 1; j < n; j++ {
+				if vn1[j] == 0 {
+					continue
+				}
+
+				// The following marked lines follow from the
+				// analysis in Lapack Working Note 176.
+				r := math.Abs(a[rk*lda+j]) / vn1[j] // *
+				temp := math.Max(0, 1-r*r)          // *
+				r = vn1[j] / vn2[j]                 // *
+				temp2 := temp * r * r               // *
+				if temp2 < tol3z {
+					vn2[j] = float64(lsticc)
+					lsticc = j
+				} else {
+					vn1[j] = vn1[j] * math.Sqrt(temp) // *
+				}
+			}
+		}
+
+		a[rk*lda+k] = akk
+	}
+	kb = k
```

Comment:
Judging from the code below (you omit +1 consistently everywhere) I think you are aware of this but anyway: when the above loop terminates, in Fortran k will point to the last processed column (which is in the matrix), while here k will point to the next one (which may be outside). Here I assume that you will move incrementing k to the end of the loop body.